### PR TITLE
Main logic flow

### PR DESCRIPTION
**Pull request by paskawonder**

### CODE REVIEW

File: src/main/java/com/epam/mentoring/library/service/LeasingHistoryService.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+    private UserRepository userRepository;
+
+    @SuppressWarnings(value = "unchecked")
+    public BookInstance borrowBook(final long userId, final long bookId, final long period) {
+        List<Long> bookInstanceIds = bookService.getBookById(bookId).getBookInstances()
+                .stream()
+                .map(BookInstance::getId).collect(Collectors.toList());
+        for (long id : bookInstanceIds) {
+            if (leasingHistoryRepository.getLeasingHistoryByBookInstanceIdAndActualEndDateIsNull(id).isEmpty()) {
```
Comment (jaworskiwoj) [inline src/main/java/com/epam/mentoring/library/service/LeasingHistoryService.java]: How we are going to handle case when user wants to boorow the same book twice (e.g. for him and his firend).

File: src/main/resources/application.properties [hunk 2]
```diff
@@ -0,0 +1,5 @@
+spring.datasource.driverClassName=org.postgresql.Driver
+spring.datasource.url=jdbc:postgresql://pg-1131be5f-paskawonder-3251.aivencloud.com:11814/defaultdb?sslmode=require
+spring.datasource.username=avnadmin
+spring.datasource.password=z9dr0cnqwplxv4bo
+spring.jpa.properties.hibernate.jdbc.lob.non_contextual_creation=true
```
Comment (jaworskiwoj) [inline src/main/resources/application.properties]: Please add unit tests for services and controllers.

File: src/main/java/com/epam/mentoring/library/model/Book.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+public final class Book {
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "book_id_seq")
+    @SequenceGenerator(name = "book_id_seq", sequenceName = "book_id_seq", allocationSize = 1)
+    private Long id;
+
+    private String title;
+
+    private String author;
```
[the remark this answers is not in the document]
Comment (paskawonder) [reply src/main/java/com/epam/mentoring/library/model/Book.java]: accepted for consideration

File: src/main/java/com/epam/mentoring/library/rest/BookRestController.java [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+import org.springframework.web.bind.annotation.RestController;
+
+import java.net.URI;
+import java.util.List;
+
+@Slf4j
+@RestController
+@RequestMapping("/books")
+@Api(value = "Book API", description = "Manipulation of book")
+@AllArgsConstructor(onConstructor = @__({@Autowired}))
```
[the remark this answers is not in the document]
Comment (paskawonder) [reply src/main/java/com/epam/mentoring/library/rest/BookRestController.java]: Is it a good practice to use field injection?

File: src/main/java/com/epam/mentoring/library/rest/UserRestController.java [hunk 5]
```diff
@@ -5,49 +5,126 @@
...
-        return null;
+    private boolean isUserDocumentIdNotUnique(final String userDocumentId) {
+        if (userService.getAllUsers().stream().anyMatch(u -> userDocumentId.equals(u.getDocumentId()))) {
+            log.info("User with the same documentId is already exist; documentId : " + userDocumentId);
+            return true;
+        }
+        return false;
     }
 
+    private boolean isUserIdExist(final long id) {
```
[the remark this answers is not in the document]
Comment (paskawonder) [reply src/main/java/com/epam/mentoring/library/rest/UserRestController.java]: accepted for consideration

File: src/main/java/com/epam/mentoring/library/model/LeasingHistory.java [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+    @ManyToOne(cascade = CascadeType.ALL)
+    @JoinColumn(name = "user_id")
+    private User user;
+
+    private LocalDate startDate;
+
+    private  LocalDate arrangedEndDate;
+
+    @Setter
```
[the remark this answers is not in the document]
Comment (paskawonder) [reply src/main/java/com/epam/mentoring/library/model/LeasingHistory.java]: Should we care about mutability in general?  In this particular case I don't think that we need to provide a facility to change an user borrowed a book, start date or book instance. Maybe arrangedEndDate could be a mutable field also.

File: src/main/java/com/epam/mentoring/library/rest/LeasingHistoryController.java [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+
+    private final LeasingHistoryService leasingHistoryService;
+
+    @ApiOperation(value = "statistics/mostPopularBooks", notes = "Most Popular Books")
+    @GetMapping("statistics/mostPopularBooks")
+    public ResponseEntity<Map<Book, Long>> mostPopularBooks() {
+        Map<Book, Long> mostPopularBooks = leasingHistoryService.mostPopularBooks();
+        return new ResponseEntity<>(mostPopularBooks, HttpStatus.OK);
+    }
+    @ApiOperation(value = "statistics/dishonestUsers", notes = "Most Popular Books")
```
[the remark this answers is not in the document]
Comment (paskawonder) [reply src/main/java/com/epam/mentoring/library/rest/LeasingHistoryController.java]: accepted for consideration

File: src/main/resources/application.properties (hunk 2 above)
Comment (paskawonder) [reply src/main/resources/application.properties]: accepted for consideration

**Pull request merged**